eta/core/datasets.py

```diff
@@ -1370,19 +1373,294 @@ def transform(self, src):
+    Balance the the dataset's values of a categorical attribute by removing
+    records.
+
+    For example:
+        Given a dataset with 10 green cars, 20 blue cars and 15 red cars,
+        remove records with blue and red cars until there are the same number
+        of each color.
+
+        In this example 'color' is the `attribute_name` and 'car' is the
+        `object_label`.
     '''
 
-    def __init__(self, attribute):
-        self.attr = attribute
+    def __init__(self, attribute_name, object_label=None,
+                 median_point=0.25, negative_power=5.0):
+        '''
+
+        Args:
+            attribute_name (str): the name of the attribute to balance by
+            object_label (str): the name of the object label that the
+                attribute_name must be nested under. If this is None, it is
+                assumed that the attributes are Image/Frame level attrs.
+            median_point (float): value between [0, 1] to specify what the
+                target count per attribute value will be.
+                0.5 - will result in the true median
+                0   - the minimum value
+                It is recommended to set this somewhere between [0, 0.5]. The
+                smaller this value is, the closer all values can be balanced,
+                at the risk that if some values have particularly low number of
+                samples, they dataset will be excessively trimmed.
+            negative_power (float): value between [1, ~LARGE~] that weights the
+                negative values (where the count of a value is less that the
+                target) when computing the score for a set of indices to remove.
+                1 - will weight them the same as positive values
+                2 - will square the values
+                ...
+                Check Balancer._solution_score for more details.
+        '''
+        self.attr_name = attribute_name
+        self.object_label = object_label
+        self.median_point = median_point
+        self.negative_power = negative_power
 
     def transform(self, src):
-        # @TODO implement Balancing!!
+        '''
+        Modify the BuilderDataset records by removing records until the
+        target attribute is ~roughly~ balanced for each value.
+
+        Args:
+            src (BuilderDataset): the dataset builder
+        '''
+
+        # STEP 1: Get attribute value(s) for every record
+        helper_list = self._to_helper_list(src.records)
+
+        # STEP 2: determine target number to remove of each attribute value
+        target_remove = self._get_target_remove(helper_list)
+
+        # STEP 3: find the records to keep
+        keep_idxs = self._get_keep_idxs(target_remove, helper_list)
+
+        # STEP 4: modify the list of records
         old_records = src.records
         src.clear()
-        pass
+        for ki in keep_idxs:
+            src.add(old_records[helper_list[ki][0]])
 
+    def _to_helper_list(self, records):
```

Comment:
maybe split this into two functions. 
helper_video
helper_image 

And call them from _to_helper_list.
This function is rather long.